flink-ml-benchmark/src/main/java/org/apache/flink/ml/benchmark/datagenerator/common/DenseVectorArrayGenerator.java

```diff
@@ -42,7 +42,7 @@ protected RowGenerator[] getRowGenerators() {
         return new RowGenerator[] {
             new RowGenerator(getNumValues(), getSeed()) {
                 @Override
-                protected Row nextRow() {
+                protected Row getRow() {
```

Comment:
Reusing one row to simulate the input may cause the benchmark result not stable, since processing two different rows may take different time.

The two optimizations introduced in this PR give us 12% speedup. Is there a breakdown analysis of this two optimization?

If the data generation cost is indeed high, can we sample a fixed number of rows and reuse it later?

Reply:
Good point. I have updated the PR as suggested. The RowGenerator would generate up to 64 rows and re-use those rows as the source function output.

I have updated the PR description with benchmark results for 3 different setups. It shows the percentage improvement provided by each optimization. The overall execution time is smaller than what I observed last time and the percentage difference is smaller.

Reply:
Thanks for the update. As generating 10^8 rows takes ~2s, do you think generating data is still a big overhead that needs to be addressed? Since reusing some of the rows does introduce some stability issues.

Reply:
I think we need to make this optimization since 2 sec still a big overhead. And previously I have seen bigger difference (more than 3 sec) after we make this optimization.

We expect Flink ML's performance to be close to Spark ML. That means we expect Flink ML's total time for this benchmark to be less than 6 sec after we finish all optimizations. Then 2 sec is going to take a big percentage of the total time.

Can you explain why this PR will introduce stability issues? Did you observe additional stability issues with this PR?

Reply:
I have updated the PR such that it will pre-generate 1/1000 of the expected number of rows and re-use those rows as the source function output. This allows the number of unique values to increase with the total number of expected values. Can you see if this address the concern?

Reply:
Thanks for the update. Setting the number of pre-generated data as 1/1000 does reduce the variance of the benchmark result here. LGTM.